src/Core/src/Credentials/ConfigurationProvider.php

```diff
@@ -13,20 +17,71 @@
  */
 final class ConfigurationProvider implements CredentialProvider
 {
+    private $logger;
+
+    private $httpClient;
+
+    public function __construct(?LoggerInterface $logger = null, ?HttpClientInterface $httpClient = null)
+    {
+        $this->logger = $logger ?? new NullLogger();
+        $this->httpClient = $httpClient;
+    }
+
     public function getCredentials(Configuration $configuration): ?Credentials
     {
-        if (
-            $configuration->has(Configuration::OPTION_ACCESS_KEY_ID)
-            && $configuration->has(Configuration::OPTION_SECRET_ACCESS_KEY)
-        ) {
-            /** @psalm-suppress PossiblyNullArgument */
-            return new Credentials(
-                $configuration->get(Configuration::OPTION_ACCESS_KEY_ID),
-                $configuration->get(Configuration::OPTION_SECRET_ACCESS_KEY),
-                $configuration->get(Configuration::OPTION_SESSION_TOKEN)
-            );
+        $accessKeyId = $configuration->get(Configuration::OPTION_ACCESS_KEY_ID);
+        $secretAccessKeyId = $configuration->get(Configuration::OPTION_SECRET_ACCESS_KEY);
+
+        if (null === $accessKeyId || null === $secretAccessKeyId) {
+            return null;
+        }
+
+        $roleArn = $configuration->get(Configuration::OPTION_ROLE_ARN);
+        if (null !== $roleArn) {
+            $region = $configuration->get(Configuration::OPTION_REGION);
+            $roleSessionName = $configuration->get(Configuration::OPTION_ROLE_SESSION_NAME);
+
+            return $this->getCredentialsFromRole($accessKeyId, $secretAccessKeyId, $region, $roleArn, $roleSessionName);
```

Comment:
Shouldn't we provide the sessionToken (when defined) here?

```php
$credentials = new Credentials(
    $configuration->get(Configuration::OPTION_ACCESS_KEY_ID),
    $configuration->get(Configuration::OPTION_SECRET_ACCESS_KEY),
    $configuration->get(Configuration::OPTION_SESSION_TOKEN)
);
if (null === $roleArn) {
   return $credentials;
}

return $this->getCredentialsFromRole($credentials, ...);

```

Reply:
I can pass the session token to the getCredentialsFromRole method if defined, and use add it ton the Credentials for StsClient.
Is it your point ?

Reply:
Pushed the update, is it better ?